reps/2023-03-10-batch-inference.md

```diff
@@ -227,28 +237,176 @@ def preprocess(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+        """Create a serializable Checkpoint from this Predictor.
+
+        Returns:
+            Checkpoint containing the state for this Predictor.
+        """
+    
+    def __getstate__(self):
+        return self._to_checkpoint()
+
+    def __setstate__(self, checkpoint):
+        # Same logic as `from_checkpoint`.
+
+    def use_callable_class(self) -> bool:
+        return True
+```
+
+3. When using Predictors with AIR Checkpoints, move the construction logic to `setup` so that it can be called within the actor.
+
+```python
+class Predictor:
+    def _to_checkpoint(self) -> Checkpoint:
+        """Create a serializable Checkpoint from this Predictor.
+
+        Returns:
+            Checkpoint containing the state for this Predictor.
+        """
+        ...
+    
+    def setup(self):
+        if self._checkpoint:
+            # Initialize the state from the Checkpoint.
+    
+    @classmethod
+    def from_checkpoint(cls, checkpoint):
+        return Predictor(_checkpoint=checkpoint)
+
+    def use_callable_class(self) -> bool:
+        return True
+```
+
+Putting this all together, the implementation for `map_batches` will look like the following
+
+```python
+def map_batches(
+        self,
+        fn: BatchUDF,
+        *,
+        batch_size: Optional[Union[int, Literal["default"]]] = "default",
+        compute: Optional[Union[str, ComputeStrategy]] = None,
+        batch_format: Optional[str] = "default",
+        zero_copy_batch: bool = False,
+        fn_args: Optional[Iterable[Any]] = None,
+        fn_kwargs: Optional[Dict[str, Any]] = None,
+        fn_constructor_args: Optional[Iterable[Any]] = None,
+        fn_constructor_kwargs: Optional[Dict[str, Any]] = None,
+        **ray_remote_args,
+    ) -> "Dataset[Any]":
+
+    if isinstance(fn, BatchMappable):
+        fn.validate_compute_strategy(compute)
+        batch_format = fn.validate_batch_format(batch_format)
+
+        if fn.use_callable_class:
+            # Put the BatchMappable in the object store once.
+            batch_mappable_ref = ray.put(fn)
+
+            class _Wrapper:
+                def __init__(self, batch_mappable_ref: ObjectRef[BatchMappable]):
+                    self.batch_mappable = ray.get(batch_mappable_ref)
+                
+                def __call__(self, *args, **kwargs):
+                    return self.batch_mappable(*args, **kwargs)
+            
+            fn = _Wrapper
+        
+        # Rest of the implementation stays the same
+        ...
+```
+
+## FAQ
+1. If I pass in a `Predictor` directly, won't initiliazation happen for every batch?
+
+No. `Predictors` will require the use of a Callable Class, and so we wrap the Predictor in a Callable class internally, like in the example above.
+
+2. If I have an AIR Checkpoint, won't this require initialization on the driver?
+
+No. If a Predictor is created from an AIR Checkpoint, we delay the initialization to a `setup` call that happens only in the 
```

Comment:
Can we make this more explicit somehow? Also, we shouldn't expose `setup` publicly right?